dll/win32/browseui/shellbrowser.cpp

```diff
@@ -3551,6 +3554,156 @@ LRESULT CShellBrowser::RelayMsgToShellView(UINT uMsg, WPARAM wParam, LPARAM lPar
     return 0;
 }
 
+static VOID DeleteUnwantedVerbItems(IContextMenu *pContextMenu, HMENU hMenu, INT idFirst, INT idLast)
+{
+    CHAR verb[32];
+    INT cItems = ::GetMenuItemCount(hMenu);
+    for (INT iMenu = cItems - 1; iMenu >= 0; --iMenu) // In reverse order, to keep the indexes
+    {
+        // Get menu ID
+        MENUITEMINFOW mii = { sizeof(mii), MIIM_ID };
+        GetMenuItemInfoW(hMenu, iMenu, TRUE, &mii);
+        if (mii.wID == 0)
+            continue;
+
+        // Get the verb from menu ID
+        verb[0] = 0;
+        HRESULT hr = pContextMenu->GetCommandString(mii.wID - idFirst, GCS_VERBA, NULL, verb, _countof(verb));
+        if (FAILED(hr))
+            continue;
+
+        if (_strcmpi(verb, "open") == 0 ||
+            _strcmpi(verb, "link") == 0 ||
+            _strcmpi(verb, "rename") == 0 ||
+            _strcmpi(verb, "delete") == 0)
+        {
+            ::DeleteMenu(hMenu, iMenu, MF_BYPOSITION);
+        }
+    }
+}
+
+static VOID DeleteDuplicateSeparators(HMENU hMenu)
+{
+    INT cItems = ::GetMenuItemCount(hMenu);
+    BOOL bPrevSep = FALSE;
+    for (INT iMenu = cItems - 1; iMenu >= 0; --iMenu) // In reverse order, to keep the indexes
+    {
+        // Get menu type
+        MENUITEMINFOW mii = { sizeof(mii), MIIM_TYPE };
+        GetMenuItemInfoW(hMenu, iMenu, TRUE, &mii);
+        if (mii.fType == MFT_SEPARATOR)
+        {
+            if (bPrevSep)
+                ::DeleteMenu(hMenu, iMenu, MF_BYPOSITION);
+            bPrevSep = TRUE;
+        }
+        else
+        {
+            bPrevSep = FALSE;
+        }
+    }
+}
+
+HRESULT CShellBrowser::ShowSysMenuContextMenu(INT xPos, INT yPos)
+{
+    HRESULT hr;
+    CComPtr<IContextMenu> pContextMenu;
+    CComPtr<IShellFolder> pFolder;
+    LPCITEMIDLIST pidlChild;
+
+    // Prepare for context menu
+    hr = SHBindToParent(fCurrentDirectoryPIDL, IID_PPV_ARG(IShellFolder, &pFolder), &pidlChild);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+    hr = pFolder->GetUIObjectOf(m_hWnd, 1, &pidlChild, IID_IContextMenu, NULL, (void **)&pContextMenu);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    // Query context menu
+#define CMDID_FIRST 0x1000
+#define CMDID_LAST  0x7000
+    UINT uFlags = CMF_NORMAL;
+    HMENU hMenu = ::CreatePopupMenu();
+    hr = pContextMenu->QueryContextMenu(hMenu, 0, CMDID_FIRST, CMDID_LAST, uFlags);
+    if (FAILED_UNEXPECTEDLY(hr))
+    {
+        ::DestroyMenu(hMenu);
+        return hr;
+    }
+
+    // Delete some menu items
+    DeleteUnwantedVerbItems(pContextMenu, hMenu, CMDID_FIRST, CMDID_LAST);
+    DeleteDuplicateSeparators(hMenu);
+    C_ASSERT(!(CMDID_FIRST <= SC_CLOSE && SC_CLOSE <= CMDID_LAST));
+#undef CMDID_FIRST
+#undef CMDID_LAST
+
+    // Get 'Close' menu item string from system menu
```

Comment:
General question: Isn't it possible instead to call GetSystemMenu first, and then remove everything from it except for the "close" item, and then append the (edited) contents of the context menu as done above? (perhaps this would mean, calling pContextMenu->QueryContextMenu() on that system menu handle? :thinking: )
If not, why?

Reply:
The original system menu is still used for when R-Clicking non-icon space of title bar.